### PR TITLE
Peer das core

### PR DESCRIPTION
**Pull request by nalepae**
Please read this PR commit by commit.

**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**
Implement https://github.com/ethereum/consensus-specs/blob/peer-das/specs/_features/eip7594/das-core.md

### CODE REVIEW

File: beacon-chain/core/peerdas/helpers.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+		sidecar := ethpb.DataColumnSidecar{
+			ColumnIndex:                  columnIndex,
+			DataColumn:                   columnBytes,
+			KzgCommitments:               blobKzgCommitments,
+			KzgProof:                     kzgProofOfColumnBytes,
+			SignedBlockHeader:            signedBlockHeader,
+			KzgCommitmentsInclusionProof: kzgCommitmentsInclusionProof,
+		}
+
+		sidecars = append(sidecars, sidecar)
```
Comment (nalepae) [inline beacon-chain/core/peerdas/helpers.go:246]: To fix: `call of append copies lock value`

File: beacon-chain/p2p/subnets.go [hunk 2]
```diff
@@ -208,10 +208,16 @@ func initializePersistentColumnSubnets(id enode.ID) error {
 	if ok && expTime.After(time.Now()) {
 		return nil
 	}
-	subs, err := computeSubscribedColumnSubnets(id)
+	subsMap, err := peerdas.CustodyColumnSubnets(id, params.BeaconConfig().CustodyRequirement)
 	if err != nil {
 		return err
 	}
+
+	subs := make([]uint64, 0, len(subsMap))
```
Comment (nalepae) [inline beacon-chain/p2p/subnets.go:216]: Ideally, I think all subnets list should be converted at all levels into subnets maps.
As far as I know:
1. We want items in these lists not to have duplicates
2. We at some places look if an item is in the list (`O(n)`)

==> These 2 conditions makes a set to be a good candidate to store those data. Moreover lookup will go from `O(n)` to `O(1)`.

File: beacon-chain/core/peerdas/helpers.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+	// Check if the custody subnet count is larger than the data column sidecar subnet count.
+	if custodySubnetCount > dataColumnSidecarSubnetCount {
+		return nil, errCustodySubnetCountTooLarge
+	}
+
+	// First, compute the subnet IDs that the node should participate in.
+	subnetIds := make(map[uint64]bool, custodySubnetCount)
+	i := uint64(0)
+
+	for uint64(len(subnetIds)) < custodySubnetCount {
```
Comment (nisdas) [inline beacon-chain/core/peerdas/helpers.go:86]: Can you make `i` part of the for loop, separating out the way it is done is odd

File: beacon-chain/p2p/subnets.go [hunk 4]
```diff
@@ -235,46 +241,6 @@ func computeSubscribedSubnets(nodeID enode.ID, epoch primitives.Epoch) ([]uint64
...
-	for _, sub := range subs {
-		for i := uint64(0); i < colsPerSub; i++ {
-			colId := params.BeaconConfig().DataColumnSidecarSubnetCount*i + sub
-			colIdxs = append(colIdxs, colId)
-		}
-	}
-	return colIdxs, nil
-}
-
-func computeSubscribedColumnSubnets(nodeID enode.ID) ([]uint64, error) {
```
Comment (nisdas) [inline beacon-chain/p2p/subnets.go:254]: please remove `computeSubscribedColumnSubnet` if the function is unused

File: beacon-chain/core/peerdas/helpers.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+		for subnetId := range subnetIds {
+			columnIndex := dataColumnSidecarSubnetCount*i + subnetId
+			columnIndices[columnIndex] = true
+		}
+	}
+
+	return columnIndices, nil
+}
+
+func CustodyColumnSubnets(nodeId enode.ID, custodySubnetCount uint64) (map[uint64]bool, error) {
```
Comment (nisdas) [inline beacon-chain/core/peerdas/helpers.go:74]:
```suggestion
func CustodyColumnSubnets(nodeId enode.ID) (map[uint64]bool, error) {
```
I do not think the custody count should be an argument, it is not dynamic but a value determined by the specification

Comment (nalepae) [reply beacon-chain/core/peerdas/helpers.go:74]: Actually the spec defines [get_custody_columns](https://github.com/ethereum/consensus-specs/blob/dev/specs/_features/eip7594/das-core.md#get_custody_columns) with the custody count as an argument:

```python
def get_custody_columns(node_id: NodeID, custody_subnet_count: uint64) -> Sequence[ColumnIndex]:
    assert custody_subnet_count <= DATA_COLUMN_SIDECAR_SUBNET_COUNT

    subnet_ids = []
    i = 0
    while len(subnet_ids) < custody_subnet_count:
        subnet_id = (
            bytes_to_uint64(hash(uint_to_bytes(uint64(node_id + i)))[0:8])
            % DATA_COLUMN_SIDECAR_SUBNET_COUNT
        )
        if subnet_id not in subnet_ids:
            subnet_ids.append(subnet_id)
        i += 1
    assert len(subnet_ids) == len(set(subnet_ids))

    columns_per_subnet = NUMBER_OF_COLUMNS // DATA_COLUMN_SIDECAR_SUBNET_COUNT
    return [
        ColumnIndex(DATA_COLUMN_SIDECAR_SUBNET_COUNT * i + subnet_id)
        for i in range(columns_per_subnet)
        for subnet_id in subnet_ids
    ]
```

The custody subnet count has an impact of the number of custody column subnet count.

>  it is not dynamic but a value determined by the specification

My understanding is the specification determines [CUSTODY_REQUIREMENT](https://github.com/ethereum/consensus-specs/blob/dev/specs/_features/eip7594/das-core.md#custody-setting)  (Minimum number of subnets an honest node custodies and serves samples from).
```go
params.BeaconConfig().CustodyRequirement // in Prysm
```

`CUSTODY_REQUIREMENT` is a **minimum** custody subnet count, but it is possible for a node to custody more. (And in the extreme case, a "supernode" can custody all the subnets.)

In my opinion, both for spec compliance and possibility of custodying more than the minimum required, it's better to keep the custody count as an argument.

File: beacon-chain/core/peerdas/helpers.go (hunk 3 above)
Comment (nalepae) [reply beacon-chain/core/peerdas/helpers.go]: Fixed in https://github.com/prysmaticlabs/prysm/pull/13877/commits/039f9eb7e950e98b12172664843dda22b0270175.

**Pull request merged**